File: ruby/lib/steam/community/css/css_stats.rb
# This code is free software; you can redistribute it and/or modify it under the
# terms of the new BSD License.
#
# Copyright (c) 2010, Sebastian Staudt

require 'steam/community/css/css_map'
require 'steam/community/css/css_weapon'

# The CSSStats class represents the game statistics for a single user in
# Counter-Strike: Source
class CSSStats

  include GameStats

  MAPS = [ 'cs_assault', 'cs_compound', 'cs_havana', 'cs_italy', 'cs_militia',
           'cs_office', 'de_aztec', 'de_cbble', 'de_chateau', 'de_dust',
           'de_dust2', 'de_inferno', 'de_nuke', 'de_piranesi', 'de_port',
           'de_prodigy', 'de_tides', 'de_train' ]

  WEAPONS = [ 'deagle', 'usp', 'glock', 'p228', 'elite', 'fiveseven', 'awp',
              'ak47', 'm4a1', 'aug', 'sg552', 'sg550', 'galil', 'famas',
              'scout', 'g3sg1', 'p90', 'mp5navy', 'tmp', 'mac10', 'ump45',
              'm3', 'xm1014', 'm249', 'knife', 'grenade' ]

  attr_reader :last_match_stats, :total_stats

  # Creates a CSSStats object by calling the super constructor with the game
  # name "cs:s"
  def initialize(steam_id)
    super steam_id, 'cs:s'

    if public?
      @last_match_stats = {}
      @total_stats      = {}

      @last_match_stats[:cost_per_kill]   = @xml_data.elements['stats/lastmatch/costkill'].text.to_f
      @last_match_stats[:ct_wins]         = @xml_data.elements['stats/lastmatch/ct_wins'].text.to_i
      @last_match_stats[:damage]          = @xml_data.elements['stats/lastmatch/dmg'].text.to_i
      @last_match_stats[:deaths]          = @xml_data.elements['stats/lastmatch/deaths'].text.to_i
      @last_match_stats[:dominations]     = @xml_data.elements['stats/lastmatch/dominations'].text.to_i
      @last_match_stats[:favorite_weapon] = @xml_data.elements['stats/lastmatch/favwpn'].text
      @last_match_stats[:kills]           = @xml_data.elements['stats/lastmatch/kills'].text.to_i
      @last_match_stats[:max_players]     = @xml_data.elements['stats/lastmatch/max_players'].text.to_i
      @last_match_stats[:money]           = @xml_data.elements['stats/lastmatch/money'].text.to_i
      @last_match_stats[:revenges]        = @xml_data.elements['stats/lastmatch/revenges'].text.to_i
      @last_match_stats[:stars]           = @xml_data.elements['stats/lastmatch/stars'].text.to_i
      @last_match_stats[:t_wins]          = @xml_data.elements['stats/lastmatch/t_wins'].text.to_i
      @last_match_stats[:wins]            = @xml_data.elements['stats/lastmatch/wins'].text.to_i
      @total_stats[:blind_kills]          = @xml_data.elements['stats/lifetime/blindkills'].text.to_i
      @total_stats[:bombs_defused]        = @xml_data.elements['stats/lifetime/bombsdefused'].text.to_i
      @total_stats[:bombs_planted]        = @xml_data.elements['stats/lifetime/bombsplanted'].text.to_i
      @total_stats[:damage]               = @xml_data.elements['stats/lifetime/dmg'].text.to_i
      @total_stats[:deaths]               = @xml_data.elements['stats/summary/deaths'].text.to_i
      @total_stats[:domination_overkills] = @xml_data.elements['stats/lifetime/dominationoverkills'].text.to_i
      @total_stats[:dominations]          = @xml_data.elements['stats/lifetime/dominations'].text.to_i
      @total_stats[:earned_money]         = @xml_data.elements['stats/lifetime/money'].text.to_i
      @total_stats[:enemy_weapon_kills]   = @xml_data.elements['stats/lifetime/enemywpnkills'].text.to_i
      @total_stats[:headshots]            = @xml_data.elements['stats/lifetime/headshots'].text.to_i
      @total_stats[:hits]                 = @xml_data.elements['stats/summary/shotshit'].text.to_i
      @total_stats[:hostages_rescued]     = @xml_data.elements['stats/lifetime/hostagesrescued'].text.to_i
      @total_stats[:kills]                = @xml_data.elements['stats/summary/kills'].text.to_i
      @total_stats[:knife_kills]          = @xml_data.elements['stats/lifetime/knifekills'].text.to_i
      @total_stats[:logos_sprayed]        = @xml_data.elements['stats/lifetime/decals'].text.to_i
      @total_stats[:nightvision_damage]   = @xml_data.elements['stats/lifetime/nvgdmg'].text.to_i
      @total_stats[:pistol_rounds_won]    = @xml_data.elements['stats/lifetime/pistolrounds'].text.to_i
      @total_stats[:revenges]             = @xml_data.elements['stats/lifetime/revenges'].text.to_i
      @total_stats[:rounds_played]        = @xml_data.elements['stats/summary/rounds'].text.to_i
      @total_stats[:rounds_won]           = @xml_data.elements['stats/summary/wins'].text.to_i
      @total_stats[:seconds_played]       = @xml_data.elements['stats/summary/timeplayed'].text.to_i
      @total_stats[:shots]                = @xml_data.elements['stats/summary/shots'].text.to_i
      @total_stats[:stars]                = @xml_data.elements['stats/summary/stars'].text.to_i
      @total_stats[:weapons_donated]      = @xml_data.elements['stats/lifetime/wpndonated'].text.to_i
      @total_stats[:windows_broken]       = @xml_data.elements['stats/lifetime/winbroken'].text.to_i
      @total_stats[:zoomed_sniper_kills]  = @xml_data.elements['stats/lifetime/zsniperkills'].text.to_i

      @last_match_stats[:kdratio] = @last_match_stats[:kills].to_f / @last_match_stats[:deaths]
      @total_stats[:accuracy]     = @total_stats[:hits].to_f / @total_stats[:shots]
      @total_stats[:kdratio]      = @total_stats[:kills].to_f / @total_stats[:deaths]
      @total_stats[:rounds_lost]  = @total_stats[:rounds_played] - @total_stats[:rounds_won]
    end
  end

  # Returns a Hash of CSSMap for this user containing all Counter-Strike:
  # Source maps. If the maps haven't been parsed already, parsing is done now.
  def map_stats
    return unless public?

    if @map_stats.nil?
      @map_stats = {}
      maps_data = @xml_data.elements['stats/maps']

      MAPS.each do |map_name|
        @map_stats[map_name] = CSSMap.new(map_name, maps_data)
      end
    end

    @map_stats
  end

  # Returns a Hash of CSSWeapon for this user containing all Counter-Strike:
  # Source weapons. If the weapons haven't been parsed already, parsing is done
  # now.
  def weapon_stats
    return unless public?

    if @weapon_stats.nil?
      @weapon_stats = {}
      weapons_data = @xml_data.elements['stats/weapons']

      WEAPONS.each do |weapon_name|
        @weapon_stats[weapon_name] = CSSWeapon.new(weapon_name, weapons_data)
      end
    end

    @weapon_stats
  end

end

File: php/lib/steam/community/css/CSSStats.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2010, Sebastian Staudt
 *
 * @author     Sebastian Staudt
 * @license    http://www.opensource.org/licenses/bsd-license.php New BSD License
 * @package    Steam Condenser (PHP)
 * @subpackage Steam Community
 */

require_once 'steam/community/css/CSSMap.php';
require_once 'steam/community/css/CSSWeapon.php';

/**
 * The CSSStats class represents the game statistics for a single user in
 * Counter-Strike: Source
 *
 * @package Steam Condenser (PHP)
 * @subpackage Steam Community
 */
class CSSStats extends GameStats {

    protected static $MAPS = array( 'cs_assault', 'cs_compound',
           'cs_havana', 'cs_italy', 'cs_militia', 'cs_office', 'de_aztec',
           'de_cbble', 'de_chateau', 'de_dust', 'de_dust2', 'de_inferno',
           'de_nuke', 'de_piranesi', 'de_port', 'de_prodigy', 'de_tides',
           'de_train' );

    protected static $WEAPONS = array( 'deagle', 'usp', 'glock', 'p228',
            'elite', 'fiveseven', 'awp', 'ak47', 'm4a1', 'aug', 'sg552',
            'sg550', 'galil', 'famas', 'scout', 'g3sg1', 'p90', 'mp5navy',
            'tmp', 'mac10', 'ump45', 'm3', 'xm1014', 'm249', 'knife',
            'grenade' );

    private $lastMatchStats;

    private $totalStats;

    /**
     * Creates a CSSStats object by calling the super constructor with the game
     * name "cs:s"
     * @param steamId The custom URL or the 64bit Steam ID of the user
     * @throws SteamCondenserException If an error occurs
     */
    public function __construct($steamId) {
        parent::__construct($steamId, 'cs:s');

        if($this->isPublic()) {
            $statsData = $this->xmlData->stats;
            $this->lastMatchStats = array();
            $this->totalStats     = array();

            $this->lastMatchStats['costPerKill'] = (float) $statsData->lastmatch->costkill;
            $this->lastMatchStats['ctWins'] = (int) $statsData->lastmatch->ct_wins;
            $this->lastMatchStats['damage'] = (int) $statsData->lastmatch->dmg;
            $this->lastMatchStats['deaths'] = (int) $statsData->lastmatch->deaths;
            $this->lastMatchStats['dominations'] = (int) $statsData->lastmatch->dominations;
            $this->lastMatchStats['favoriteWeapon'] = (string) $statsData->lastmatch->favwpn;
            $this->lastMatchStats['kills'] = (int) $statsData->lastmatch->kills;
            $this->lastMatchStats['maxPlayers'] = (int) $statsData->lastmatch->max_players;
            $this->lastMatchStats['money'] = (int) $statsData->lastmatch->money;
            $this->lastMatchStats['revenges'] = (int) $statsData->lastmatch->revenges;
            $this->lastMatchStats['stars'] = (int) $statsData->lastmatch->stars;
            $this->lastMatchStats['tWins'] = (int) $statsData->lastmatch->t_wins;
            $this->lastMatchStats['wins'] = (int) $statsData->lastmatch->wins;
            $this->totalStats['blindKills'] = (int) $statsData->lifetime->blindkills;
            $this->totalStats['bombsDefused'] = (int) $statsData->lifetime->bombsdefused;
            $this->totalStats['bombsPlanted'] = (int) $statsData->lifetime->bombsplanted;
            $this->totalStats['damage'] = (int) $statsData->lifetime->dmg;
            $this->totalStats['deaths'] = (int) $statsData->summary->deaths;
            $this->totalStats['dominationOverkills'] = (int) $statsData->lifetime->dominationoverkills;
            $this->totalStats['dominations'] = (int) $statsData->lifetime->dominations;
            $this->totalStats['earnedMoney'] = (int) $statsData->lifetime->money;
            $this->totalStats['enemyWeaponKills'] = (int) $statsData->lifetime->enemywpnkills;
            $this->totalStats['headshots'] = (int) $statsData->lifetime->headshots;
            $this->totalStats['hits'] = (int) $statsData->summary->shotshit;
            $this->totalStats['hostagesRescued'] = (int) $statsData->lifetime->hostagesrescued;
            $this->totalStats['kills'] = (int) $statsData->summary->kills;
            $this->totalStats['knifeKills'] = (int) $statsData->lifetime->knifekills;
            $this->totalStats['logosSprayed'] = (int) $statsData->lifetime->decals;
            $this->totalStats['nightvisionDamage'] = (int) $statsData->lifetime->nvgdmg;
            $this->totalStats['pistolRoundsWon'] = (int) $statsData->lifetime->pistolrounds;
            $this->totalStats['revenges'] = (int) $statsData->lifetime->revenges;
            $this->totalStats['roundsPlayed'] = (int) $statsData->summary->rounds;
            $this->totalStats['roundsWon'] = (int) $statsData->summary->wins;
            $this->totalStats['secondsPlayed'] = (int) $statsData->summary->timeplayed;
            $this->totalStats['shots'] = (int) $statsData->summary->shots;
            $this->totalStats['stars'] = (int) $statsData->summary->stars;
            $this->totalStats['weaponsDonated'] = (int) $statsData->lifetime->wpndonated;
            $this->totalStats['windowsBroken'] = (int) $statsData->lifetime->winbroken;
            $this->totalStats['zoomedSniperKills'] = (int) $statsData->lifetime->zsniperkills;

            $this->lastMatchStats['kdratio'] = ($this->totalStats['deaths'] > 0) ? $this->lastMatchStats['kills'] / $this->lastMatchStats['deaths'] : 0;
            $this->totalStats['accuracy'] = ($this->totalStats['shots'] > 0) ? $this->totalStats['hits'] / $this->totalStats['shots'] : 0;
            $this->totalStats['kdratio'] = ($this->totalStats['deaths'] > 0) ? $this->totalStats['kills'] / $this->totalStats['deaths'] : 0;
            $this->totalStats['roundsLost'] = $this->totalStats['roundsPlayed'] - $this->totalStats['roundsWon'];
        }
    }

    public function getLastMatchStats() {
        return $this->lastMatchStats;
    }

    public function getMapStats() {
        if(!$this->isPublic()) {
            return null;
        }

        if($this->mapStats == null) {
            $this->mapStats = array();
            $mapsData = $this->xmlData->stats->maps;

            foreach(self::$MAPS as $mapName) {
                $this->mapStats[$mapName] = new CSSMap($mapName, $mapsData);
            }
        }

        return $this->mapStats;
    }

    public function getTotalStats() {
        return $this->totalStats;
    }

    public function getWeaponStats() {
        if(!$this->isPublic()) {
            return null;
        }

        if($this->weaponStats == null) {
            $this->weaponStats = array();
            $weaponData = $this->xmlData->stats->weapons;

            foreach(self::$WEAPONS as $weaponName) {
                $this->weaponStats[$weaponName] = new CSSWeapon($weaponName, $weaponData);
            }
        }

        return $this->weaponStats;
    }

}
?>
